crates/typst-eval/src/code.rs

```diff
@@ -54,7 +54,17 @@ fn eval_code<'a>(
 
         output = ops::join(output, value).at(span)?;
 
-        if vm.flow.is_some() {
+        if let Some(event) = &vm.flow {
+            if matches!(output, Value::Content(_)) {
+                if let FlowEvent::Return(span, Some(_), false) = event {
+                    vm.engine.sink.warn(warning!(
+                        *span,
+                        "unconditional explicit return value discards content";
+                        hint: "use `return` without a value to return the joined content"
```

Comment:
that could be the fix or no return at all. tbh, I think the most common case is returning a _non-content_ value while also trying to update a counter. 

@PgBiel do you have some thoughts on the best hint(s)?

Reply:
actually I didn't even know `return` automatically returns the joined values. :sweat_smile:

For starters, there should be a hint if there is a state update or counter update among the joined content, mentioning that a state update (or counter update / step) needs to be returned and placed in the document to be effective.

Then, another potential hint could be to "try omitting the `return` keyword to also join this value" if it is also content.